### ISSUE TITLE
Expose OCINumber (dpiNumber)

### ISSUE THREAD

**Issue by felipenoris** [oracle/odpi#100]
See commit at branch https://github.com/felipenoris/odpi/tree/fn/numeric.

# Read and Write of `dpiNumber` values

The following changes allows odpi users to bind and read `dpiNumber` from a variable

* native type `DPI_NATIVE_TYPE_NUMERIC`

* `dpiNumber` struct with 22 bytes

* `asNumber` variant for `dpiDataBuffer`. It does not increase `sizeof(dpiDataBuffer)`.

* `dpiData_getNumber` and `dpiData_setNumber`

* `dpiDataBuffer__fromOracleNumberAsNumber`

* `dpiDataBuffer__toOracleNumberFromNumber`

* `dpiOci__numberAssign` wraps `OCINumberAssign`

* `dpiVar` now reads and writes data of native type `DPI_NATIVE_TYPE_NUMERIC`.

# `dpiNumber` arithmetic

See code at the new file `dpiNumber.c`. I defined `dpiNumber_add`  and `dpiNumber_abs`. But, for some reason, I get a segfault when using them from Julia.

I had no problems on `dpiOci__numberAssign`, but it handles values from odpi to oci.
When I feed an external `dpiNumber` to `dpiNumber_add` and `dpiNumber_abs` it segfaults.
I tried to debug the value that odpi receives, and all bytes are correct. So I don't know what I'm doing wrong. Can you help me on this one?

Ref #4 .

**Comment by anthony-tuininga** [comment]
I will take a look later when I have a bit more time available.

**Comment by felipenoris** [comment]
If it would make things easier, I can isolate the arithmetic issue in a separate commit, since this is the only non-working portion of this development.

**Comment by tgulacsi** [comment]
I'm a C noob, but can you use a pointer to some value-received structs?
In "dpiNumber_add(const dpiNumber number1, const dpiNumber number2, dpiNumber *result)" but use it as "dpiOci__numberAdd(&error, &number1, &number2, result)".

Why not *dpiNumber all the way up?

**Comment by felipenoris** [comment]
Yes, that was my first attempt btw. I'll just revert to that and separate the commits.

**Comment by felipenoris** [comment]
So, for the sake of simplicity, I removed all non-working parts of the commit at https://github.com/felipenoris/odpi/tree/fn/numeric, which are related to dpiNumber arithmetic.

With this, the patch will only expose a `dpiNumber`.

**Comment by felipenoris** [comment]
@anthony-tuininga , please let me know if you need anything else on this.

**Comment by anthony-tuininga** [comment]
@felipenoris, I don't think I need anything else. Thanks. I'm not sure when this will be implemented. As you're likely aware there are other priorities!

**Comment by acautin** [comment]
@anthony-tuininga this would be helpful as for me as I am trying to work with a database explorer which uses erlang/odpi as the driver in erlang. Precision is lost on the way out and I am trying to understand if this is a limitation of `odpi` or the erlang driver on top of it.

Do you have an example reading numeric values similar to `SELECT 2.700000000000005 FROM dual` with full precision ?. (text or otherwise).

**Comment by anthony-tuininga** [comment]
@acautin, you can tell ODPI-C to fetch the data as a string representation of the OCI number instead of a double precision floating point number (the default). You can see this happening in the [sample](https://github.com/oracle/odpi/blob/master/samples/DemoFetch.c). If you create variables instead (by calling [dpiConn_newVar()](https://oracle.github.io/odpi/doc/functions/dpiConn.html#c.dpiConn_newVar)), then make sure the ``nativeTypeNum`` parameter is set to ``DPI_NATIVE_TYPE_BYTES``. Either of these methods will return a string representation of the OCI decimal number without any loss of precision.

**Comment by felipenoris** [comment]
Hi @anthony-tuininga ! Long time!
I would like to ask if it is possible for you to take a look at this.
See commit https://github.com/felipenoris/odpi/commit/3204a058ab5416b776d936c2a87d24a9d99f905b . The idea is to just expose a `dpiNumber` struct.

**Comment by anthony-tuininga** [comment]
Yes, it has been a while! The commit you reference seems to be simply exposing the 22 bytes of data -- which are in a format that is not easily consumed. How do you intend to consume and generate those 22 bytes of data?

**Comment by felipenoris** [comment]
@anthony-tuininga thanks for the reply!

The idea is to let driver developers to access the raw 22 bytes of OraNumber, instead of having to rely on the string representation to get the exact fixed decimal representation.

Obviously the driver developer has to parse these 22 bytes.

I'm already doing this in [Oracle.jl](https://github.com/felipenoris/Oracle.jl). These bytes are loaded into this 22-bytes struct:

```julia
struct OraNumber <: Real
    len::UInt8
    ex::UInt8
    mantissa::NTuple{MAX_SIZEOF_MANTISSA, UInt8}
end
```

From that, I can either implement an [API around this struct](https://github.com/felipenoris/Oracle.jl/tree/master/src/oranumbers), or translate to other fixed decimal format.

Hopefully, using these 22-bytes directly is faster than converting from a string representation.

**Comment by anthony-tuininga** [comment]
One possibility that might make sense: instead of targeting OCINumber directly, I could add a new "native type" called ``DPI_NATIVE_TYPE_OCI`` which simply provides access to the OCI value directly. You can use it for OCINumber but it could be used for other types as well in the future, if that proves useful. I might also expose methods to convert OCINumber to double and text, simply for testing purposes. I believe that should work for you but comments are welcome!

**Comment by felipenoris** [comment]
@anthony-tuininga, yes, I think that works, as long as it returns a copy of the data and not a volatile reference in memory. The `dpiNumber` struct I created was just an auxiliary struct for transferring OCINumber to/from ODPI-C.

**Comment by anthony-tuininga** [comment]
I'm not sure what you mean by "volatile reference in memory". The 22 bytes will be stored in the variable buffers and dpiDataBuffer.asOci will be a pointer to that buffer. I believe that's what you meant but can you confirm?

**Comment by anthony-tuininga** [comment]
I'm not sure what you mean by "volatile reference in memory". The 22 bytes will be stored in the variable buffers and dpiDataBuffer.asOci will be a pointer to that buffer. I believe that's what you meant but can you confirm?

**Comment by felipenoris** [comment]
@anthony-tuininga , yes I think that works. I'm sure you understand the internals of OCI better than I do. I had to rely on `OCINumberAssign` to copy the data in a `dpiData` field so that I could access the data (also setting isNull information on `dpiData`). I understand that what you're suggesting is a more general approach.

**Comment by anthony-tuininga** [comment]
> If I select the native type DPI_NATIVE_TYPE_OCI how will I know that it will return a OCINumber and not other data structure? (possible answer: by combining with the information on the Oracle Type + Native Type) ?

Yes, you would look at the Oracle Type to determine what actual value you are examining.

**Comment by felipenoris** [comment]
@anthony-tuininga that would be awesome!

**Comment by tgulacsi** [comment]
Hi,

Go's database/sql seems to get support for decimal handling: https://github.com/golang/go/issues/30870#issuecomment-717358078 

Which means I'm interested in getting/setting OCINumbers directly.

**Comment by cjbj** [comment]
> Go's database/sql seems to get support for decimal handling: [golang/go#30870 (comment)](https://github.com/golang/go/issues/30870#issuecomment-717358078)

Neat!
